File: java/leetcode/editor/cn/CombinationSumIi.java
package leetcode.editor.cn;

import java.util.*;

//[40] 组合总和 II
/* [10,1,2,7,6,1,5],8 */
public class CombinationSumIi {
    public static void main(String[] args) {
        Solution solution = new CombinationSumIi().new Solution();
    }

    //leetcode submit region begin(Prohibit modification and deletion)
    class Solution {
        public List<List<Integer>> combinationSum2(int[] candidates, int target) {
            Arrays.sort(candidates);
            List<List<Integer>> result = new ArrayList<>();
            backtrack(candidates, target, result, 0, new ArrayDeque<Integer>());
            return result;
        }

        private void backtrack(int[] candidates, int target, List<List<Integer>> result, int begin, Deque<Integer> path) {
            if (target == 0) {
                result.add(new ArrayList<>(path));
                return;
            }
            for (int i = begin; i < candidates.length; i++) {
                int current = candidates[i];
                if (target < current) {
                    break;
                }

                //去重
                if (i > begin && current == candidates[i - 1]) {
                    continue;
                }
                path.addLast(current);
                backtrack(candidates, target - current, result, i + 1, path);
                path.removeLast();
            }
        }
    }
//leetcode submit region end(Prohibit modification and deletion)

}

File: src/main/java/leetcode/editor/cn/CombinationSumIi.kt
package leetcode.editor.cn

import org.hamcrest.Matcher
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.*
import util.list2DMatcher
import util.to1DIntArray
import util.to2DIntList
import java.util.*
import kotlin.collections.ArrayList
import kotlin.collections.HashMap


// 40 组合总和 II
/* [10,1,2,7,6,1,5],8 */

fun main() {
    val result = CombinationSumIi.Solution()::combinationSum2
    assertThat(result("[10,1,2,7,6,1,5]".to1DIntArray(), 8), list2DMatcher("[[1,7],[1,2,5],[2,6],[1,1,6]]".to2DIntList()))
    assertThat(result("[2,5,2,1,2]".to1DIntArray(), 5), list2DMatcher("[[1,2,2],[5]]".to2DIntList()))
}

class CombinationSumIi {
    //leetcode submit region begin(Prohibit modification and deletion)
    class Solution {
        fun combinationSum2(candidates: IntArray, target: Int): List<List<Int>> {
            // 根据count计数的最小堆
            val frequency = PriorityQueue<Pair<Int, Int>> { a, b -> if (a.second == b.second) b.first - a.first else a.second - b.second }
            val fMap = HashMap<Int, Int>()
            candidates.forEach { i -> fMap.compute(i) { _, v -> if (v == null) 1 else v + 1 } }
            fMap.forEach { (k, v) -> frequency.offer(k to v) }
            val visit = frequency.toList()

            val result = ArrayList<List<Int>>()
            val temp = LinkedList<Int>()
            fun dfs(remind: Int, idx: Int) {
                when {
                    remind == 0 -> result.add(ArrayList(temp))
                    remind < 0 || idx >= visit.size -> return
                    else -> {
                        val pair = visit[idx]
                        for (time in 0..pair.second) {
                            repeat(time) { temp.addLast(pair.first) }
                            dfs(remind - time * pair.first, idx + 1)
                            repeat(time) { temp.removeLast() }
                        }
                    }
                }
            }
            dfs(target, 0)
            return result
        }
    }
//leetcode submit region end(Prohibit modification and deletion)

}